src/optimization/steepest_descent_method.cpp

```diff
@@ -0,0 +1,292 @@
+    vector<double> next_right_border_point(init_parameters.size());
+    double left_border = 0;
+    double next_left_border = 0;
+    double next_right_border = 0;
+    double constriction_number = 0.382;
+    double right_border = step_for_borders;
+	
+    for (size_t i = 0; i < init_parameters.size(); ++i)
+    {
+        right_border_parameters[i] = init_parameters[i] + step_for_borders*direction[i];
+    }
+	
+    next_left_border = left_border + constriction_number*(right_border - left_border);
+    next_right_border = right_border - constriction_number*(right_border - left_border);
+
+    double last_left_border = 0;
+    double last_right_border = 0;
+    double next_left_border_value = 0;
+    double next_right_border_value = 0;
+    do
+    {
+        for (size_t i = 0; i < init_parameters.size(); ++i)
+        {
+            next_left_border_point[i] = init_parameters[i] + next_left_border*direction[i];
+            next_right_border_point[i] = init_parameters[i] + next_right_border*direction[i];
+        }
+        next_left_border_value = CountObjectiveFunctionValueToMinimize(next_left_border_point);
+        next_right_border_value = CountObjectiveFunctionValueToMinimize(next_right_border_point);
+
+        last_left_border = left_border;
+        last_right_border = right_border;
+
+        if (next_left_border < next_right_border)
+        {
+            if (next_left_border_value > next_right_border_value)
+            {
+                left_border = next_left_border;
+                next_left_border = next_right_border;
+                next_right_border = right_border - constriction_number * (right_border - left_border);
+            }
+            else
+            {
+                right_border = next_right_border;
+                next_right_border = next_left_border;
+                next_left_border = left_border + constriction_number * (right_border - left_border);
+            }
+        }
+        else
+        {
+            if (next_right_border_value > next_left_border_value)
+            {
+                left_border = next_right_border;
+                next_right_border = next_left_border;
+                next_left_border = right_border - constriction_number * (right_border - left_border);
+            }
+            else
+            {
+                right_border = next_left_border;
+                next_left_border = next_right_border;
+                next_right_border = left_border + constriction_number * (right_border - left_border);
+            }
+        }
+        if ((left_border == last_left_border && right_border == last_right_border)) {
+            break;
+        }
+
+    } while ((abs(left_border - right_border) > precision_));
+
+    return (left_border + right_border) / 2;
+}
+
+double SteepestDescentMethod::Localization(
+    const vector<double>& init_parameters, 
+    const double& init_step, 
+    const vector<double>& direction) const
+{
```

Comment:
The purpose of this method is not clear. Add meaningful comment to each line explaining why to do so (not what is done) or provide exact link to an algorithm